File: php/examples/workflow_with_wasm_plugins.php
<?php

require_once __DIR__ . '/../vendor/autoload.php';

use NativeBPM\Client\Builder\Workflow;

echo "=== NativeBPM PHP SDK: Workflow with Guest WASM Plugins ===\n";

// 1. Build workflow dynamically using Workflow as Code Fluent API
echo "🔨 Building workflow dynamically using Fluent API...\n";
$workflow = new Workflow("wasm-demo", "Workflow with Guest WASM Plugins");

// Chain starting from start event
$workflow->startEvent("start")
    ->next("calculate");

// Configure WASM Service Task and chain to the AI task
$workflow->serviceTask("calculate", "Calculate Totals", "payment_topic")
    ->wasm("./calculate_total.wasm")
    ->next("aiCheck");

// Configure AI task and chain to the exclusive gateway
$workflow->aiTask("aiCheck", "AI Fraud Guard")
    ->provider("google")
    ->model("gemini-2.5-flash")
    ->prompt('Analyze transaction for fraud: ${orderAmount}')
    ->resultVar("isFraudulent")
    ->next("gateway");

// Set up exclusive gateway with transition conditions and default flow
$workflow->exclusiveGateway("gateway", "Fraud Gateway")
    ->condition("userTask", '${isFraudulent == true}')
    ->defaultValue("end");

// Configure User Task and chain to the end event
$workflow->userTask("userTask", "Manual Fraud Approval")
    ->assignee("security_officer")
    ->next("end");

$workflow->endEvent("end", "Process Finished");

$bpmnXml = $workflow->buildXML();
echo "✓ Successfully compiled WASM workflow AST to BPMN 2.0 XML.\n";

// Print a snippet of generated XML
if (strlen($bpmnXml) > 300) {
    echo "XML snippet:\n" . substr($bpmnXml, 0, 300) . "...\n";
} else {
    echo "XML output:\n" . $bpmnXml . "\n";
}

File: rust/examples/workflow_with_wasm_plugins.rs
use nativebpm_client::builder::Workflow;
use nativebpm_client::apis::{configuration, default_api};

#[tokio::main]
async fn main() {
    println!("=== NativeBPM Rust SDK: Workflow with Guest WASM Plugins ===");

    // 1. Build workflow as code using Fluent API method chaining
    let mut workflow = Workflow::new("wasm-demo", "Workflow with Guest WASM Plugins");

    // Chain starting from the start event
    workflow.start_event("start")
        .connect_to("calculate")
        .service_task("calculate", "Calculate Totals", "payment_topic")
        .wasm("./calculate_total.wasm")
        .connect_to("aiCheck")
        .ai_task("aiCheck", "AI Fraud Guard")
        .provider("google")
        .model("gemini-2.5-flash")
        .prompt("Analyze transaction for fraud: ${orderAmount}")
        .result_var("isFraudulent")
        .connect_to("gateway")
        .exclusive_gateway("gateway", "Fraud Gateway")
        .condition("userTask", "${isFraudulent == true}")
        .default("end")
        .builder()
        .user_task("userTask", "Manual Fraud Approval")
        .assignee("security_officer")
        .connect_to("end")
        .end_event("end", "Process Finished");

    // Compile the workflow AST to standard BPMN 2.0 XML using the embedded Go engine
    let bpmn_xml = match workflow.build_xml() {
        Ok(xml) => {
            println!("✓ Successfully compiled WASM workflow AST to BPMN 2.0 XML.");
            xml
        }
        Err(e) => {
            eprintln!("✗ Failed to compile workflow: {}", e);
            return;
        }
    };

    // 2. Deploy and start process definition using the REST API client
    let mut config = configuration::Configuration::new();
    config.base_path = "http://localhost:8080".to_string();
    config.bearer_access_token = Some("test-bearer-token".to_string());

    println!("\nDeploying to NativeBPM engine...");
    
    let bpmn_path = std::env::temp_dir().join("wasm-demo.bpmn");
    if std::fs::write(&bpmn_path, bpmn_xml).is_ok() {
        match default_api::deploy_definition(&config, Some(bpmn_path)).await {
            Ok(definition) => {
                println!("✓ Deployed process definition (hash: {:?})", definition.hash);

                // Start a process instance
                let mut variables = std::collections::HashMap::new();
                variables.insert("orderAmount".to_string(), serde_json::json!(2500));

                let start_request = nativebpm_client::models::StartInstanceRequest {
                    instance_id: None,
                    business_key: Some("tx-8837".to_string()),
                    variables: Some(variables),
                };

                match default_api::start_instance(&config, "wasm-demo", Some(start_request)).await {
                    Ok(instance) => {
                        println!("✓ Started process instance ID: {} (completed: {})", instance.id, instance.completed);
                    }
                    Err(e) => {
                        println!("Note: Failed to start process instance: {:?}", e);
                    }
                }
            }
            Err(e) => {
                println!("Note: Local API Engine deployment skipped (ensure local server is running on :8080). Details: {:?}", e);
            }
        }
    }
}
